Special files in a repository cannot be `pants_ignore`d

**Describe the bug**
running pants tailor on a repo with a socket file causes the following crash:
```
bash-5.1$ ls
pants		pants.toml
bash-5.1$ python -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('test.sock')"
bash-5.1$ ./pants tailor
08:27:39.96 [INFO] Initializing scheduler...
08:27:40.32 [INFO] Scheduler initialized.
08:27:40.35 [ERROR] Exception caught: (pants.engine.internals.scheduler.ExecutionError)
  File "/Users/yoav/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.4xKIPa/install/lib/python3.9/site-packages/pants/bin/local_pants_runner.py", line 234, in _run_inner
    return self._perform_run(goals)
  File "/Users/yoav/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.4xKIPa/install/lib/python3.9/site-packages/pants/bin/local_pants_runner.py", line 173, in _perform_run
    return self._perform_run_body(goals, poll=False)
  File "/Users/yoav/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.4xKIPa/install/lib/python3.9/site-packages/pants/bin/local_pants_runner.py", line 190, in _perform_run_body
    return self.graph_session.run_goal_rules(
  File "/Users/yoav/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.4xKIPa/install/lib/python3.9/site-packages/pants/init/engine_initializer.py", line 135, in run_goal_rules
    exit_code = self.scheduler_session.run_goal_rule(
  File "/Users/yoav/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.4xKIPa/install/lib/python3.9/site-packages/pants/engine/internals/scheduler.py", line 533, in run_goal_rule
    self._raise_on_error([t for _, t in throws])
  File "/Users/yoav/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.4xKIPa/install/lib/python3.9/site-packages/pants/engine/internals/scheduler.py", line 501, in _raise_on_error
    raise ExecutionError(

Exception message: 1 Exception encountered:

  Exception: Failed to scan directory "/private/tmp/zxcv/": Expected File, Dir or Link, but "test.sock" (relative to "/private/tmp/zxcv") was a FileType(FileType { mode: 49152 })



(Use --print-stacktrace for more error details and/or --no-process-execution-local-cleanup to inspect chroots and/or -ldebug for more logs. See https://www.pantsbuild.org/v2.6/docs/troubleshooting for common issues. Consider reaching out for help: https://www.pantsbuild.org/v2.6/docs/getting-help.)
```

**Pants version**
Which version of Pants are you using? 2.7

**OS**
Are you encountering the bug on MacOS, Linux, or both? both

Note that `pants_ignore` doesn't work.

```
❯ python -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('test.sock')"
❯ ./pants --pants-ignore=test.sock list :
10:23:05.45 [ERROR] 1 Exception encountered:

Engine traceback:
  in select
  in pants.backend.project_info.list_targets.list_targets
  in pants.engine.internals.graph.resolve_addresses_from_specs
  in pants.engine.internals.build_files.addresses_from_address_specs
  in paths
Traceback (no traceback):
  <pants native internals>
Exception: Failed to scan directory "/Users/ericarellano/code/pants/": Expected File, Dir or Link, but "test.sock" (relative to "/Users/ericarellano/code/pants") was a FileType(FileType { mode: 49152 })
```

--- a/src/python/pants/engine/fs_test.py
+++ b/src/python/pants/engine/fs_test.py
@@ -4,6 +4,7 @@ import hashlib
 import os
 import pkgutil
 import shutil
+import socket
 import ssl
 import tarfile
 import time
@@ -283,6 +284,22 @@ def test_path_globs_ignore_pattern(rule_runner: RuleRunner) -> None:
         ["**", "!*.ln"],
         expected_files=["4.txt", "a/3.txt", "a/b/1.txt", "a/b/2"],
         expected_dirs=["a", "a/b"],
+    )
+
+
+def test_path_globs_ignore_sock(rule_runner: RuleRunner) -> None:
+    sock_path = os.path.join(rule_runner.build_root, "sock.sock")
+    with socket.socket(socket.AF_UNIX) as sock:
+        sock.bind(sock_path)
+    assert os.path.exists(sock_path)
+    assert not os.path.isfile(sock_path)
+
+    rule_runner.write_files({"non-sock.txt": ""})
+    assert_path_globs(
+        rule_runner,
+        ["**"],
+        expected_files=["non-sock.txt"],
+        expected_dirs=[],
     )
 
 

--- a/src/rust/engine/fs/src/lib.rs
+++ b/src/rust/engine/fs/src/lib.rs
@@ -504,14 +504,14 @@ impl PosixFS {
           compute_metadata,
         )
       })
-      .filter(|s| match s {
-        Ok(ref s) =>
-        // It would be nice to be able to ignore paths before stat'ing them, but in order to apply
-        // git-style ignore patterns, we need to know whether a path represents a directory.
-        {
-          !self.ignore.is_ignored(s)
+      .filter_map(|s| match s {
+        Ok(Some(s)) if !self.ignore.is_ignored(&s) => {
+          // It would be nice to be able to ignore paths before stat'ing them, but in order to apply
+          // git-style ignore patterns, we need to know whether a path represents a directory.
+          Some(Ok(s))
         }
-        Err(_) => true,
+        Ok(_) => None,
+        Err(e) => Some(Err(e)),
       })
       .collect::<Result<Vec<_>, io::Error>>()
       .map_err(|e| {
@@ -520,7 +520,6 @@ impl PosixFS {
           format!("Failed to scan directory {:?}: {}", dir_abs, e),
         )
       })?;
-    #[allow(clippy::unnecessary_sort_by)]
     stats.sort_by(|s1, s2| s1.path().cmp(s2.path()));
     Ok(DirectoryListing(stats))
   }
@@ -581,7 +580,7 @@ impl PosixFS {
     path_for_stat: PathBuf,
     file_type: std::fs::FileType,
     compute_metadata: F,
-  ) -> Result<Stat, io::Error>
+  ) -> Result<Option<Stat>, io::Error>
   where
     F: FnOnce() -> Result<std::fs::Metadata, io::Error>,
   {
@@ -605,23 +604,17 @@ impl PosixFS {
       ));
     }
     if file_type.is_symlink() {
-      Ok(Stat::Link(Link(path_for_stat)))
+      Ok(Some(Stat::Link(Link(path_for_stat))))
     } else if file_type.is_file() {
       let is_executable = compute_metadata()?.permissions().mode() & 0o100 == 0o100;
-      Ok(Stat::File(File {
+      Ok(Some(Stat::File(File {
         path: path_for_stat,
         is_executable: is_executable,
-      }))
+      })))
     } else if file_type.is_dir() {
-      Ok(Stat::Dir(Dir(path_for_stat)))
+      Ok(Some(Stat::Dir(Dir(path_for_stat))))
     } else {
-      Err(io::Error::new(
-        io::ErrorKind::InvalidData,
-        format!(
-          "Expected File, Dir or Link, but {:?} (relative to {:?}) was a {:?}",
-          path_for_stat, absolute_path_to_root, file_type
-        ),
-      ))
+      Ok(None)
     }
   }
 
@@ -631,18 +624,16 @@ impl PosixFS {
       SymlinkBehavior::Aware => fs::symlink_metadata(abs_path),
       SymlinkBehavior::Oblivious => fs::metadata(abs_path),
     };
-    let stat_result = metadata.and_then(|metadata| {
-      PosixFS::stat_internal(&self.root.0, relative_path, metadata.file_type(), || {
-        Ok(metadata)
+    metadata
+      .and_then(|metadata| {
+        PosixFS::stat_internal(&self.root.0, relative_path, metadata.file_type(), || {
+          Ok(metadata)
+        })
       })
-    });
-    match stat_result {
-      Ok(v) => Ok(Some(v)),
-      Err(err) => match err.kind() {
+      .or_else(|err| match err.kind() {
         io::ErrorKind::NotFound => Ok(None),
         _ => Err(err),
-      },
-    }
+      })
   }
 }
 

--- a/src/rust/engine/fs/src/posixfs_tests.rs
+++ b/src/rust/engine/fs/src/posixfs_tests.rs
@@ -115,9 +115,12 @@ async fn stat_symlink_oblivious() {
 
 #[tokio::test]
 async fn stat_other() {
-  new_posixfs("/dev")
-    .stat_sync(PathBuf::from("null"))
-    .expect_err("Want error");
+  assert_eq!(
+    new_posixfs("/dev")
+      .stat_sync(PathBuf::from("null"))
+      .unwrap(),
+    None,
+  );
 }
 
 #[tokio::test]

